File: forge-core/internal/bootstrapgrantauthority/grant.go
package bootstrapgrantauthority

import (
	"fmt"
	"math"

	"forgeos/forge-core/internal/capabilitygrantcontract"
)

// Grant is a fully validated and Ed25519-authenticated bootstrap Grant.
type Grant struct{ document map[string]any }

// IssueGrant creates and authenticates the only allowed bootstrap Grant.
func IssueGrant(policy *Policy, request *Request, storedAt int64, issuer *Issuer) (*Grant, error) {
	if policy == nil || request == nil || issuer == nil || issuer.trust == nil {
		return nil, fmt.Errorf("Policy, Request, and issuer are required")
	}
	if err := validateSigningInputs(policy, request, issuer.trust); err != nil {
		return nil, err
	}
	if policy.document["disposition"] != "allow" {
		return nil, fmt.Errorf("deny Policy cannot issue a Grant")
	}
	if err := validateIssuanceTime(policy.document, request.document, storedAt); err != nil {
		return nil, err
	}
	candidate, err := buildGrantCandidate(policy.document, request.document, storedAt, issuer.trust)
	if err != nil {
		return nil, err
	}
	prepared, digest, err := capabilitygrantcontract.PrepareGrantForSigning(candidate)
	if err != nil {
		return nil, err
	}
	proof, err := issuer.sign(grantSignatureDomain, digest)
	if err != nil {
		return nil, err
	}
	document, err := capabilitygrantcontract.FinalizeSignedGrant(prepared, proof)
	if err != nil {
		return nil, err
	}
	grant := &Grant{document: document}
	if err = validateGrantRelations(grant, policy, request, storedAt, issuer.trust); err != nil {
		return nil, err
	}
	return grant, nil
}

func validateSigningInputs(policy *Policy, request *Request, trust *Trust) error {
	if err := validateAuthorityBinding(policy.document, trust, "Policy"); err != nil {
		return err
	}
	if err := validateAuthorityBinding(request.document, trust, "Request"); err != nil {
		return err
	}
	return validatePolicyRequest(policy.document, request.document)
}

func buildGrantCandidate(policy, request map[string]any, storedAt int64,
	trust *Trust) (map[string]any, error) {
	ttl, _ := intValue(request, "requested_ttl_ms")
	if ttl > math.MaxInt64-storedAt {
		return nil, fmt.Errorf("Grant expiration overflows signed int64")
	}
	requestBindings, _ := objectValue(request, "bindings")
	return map[string]any{
		"api_version": "forgeos.capability-grant/v1", "approval_refs": []any{},
		"authority_proof": buildGrantProof(trust), "bindings": map[string]any{
			"context_sha256":       requestBindings["context_sha256"],
			"grant_request_sha256": request["request_sha256"], "impact_sha256": nil,
			"plan_sha256": nil, "policy_sha256": policy["policy_sha256"], "risk_sha256": nil,
			"source_revision":    requestBindings["source_revision"],
			"source_tree_sha256": requestBindings["source_tree_sha256"],
		},
		"budget": cloneNode(request["budget"]), "canonicalization": canonicalization,
		"capability":               cloneNode(request["capability"]),
		"effect_vocabulary_sha256": effectVocabularySHA256, "grant_id": "", "grant_sha256": "",
		"issuance_phase": "bootstrap_planning", "kind": "CapabilityGrant",
		"scope": cloneNode(request["scope"]), "separation_of_duty": buildSeparation(trust),
		"subject": cloneNode(request["subject"]), "task_binding": cloneNode(request["task_binding"]),
		"usage_policy": buildUsagePolicy(), "validity": map[string]any{
			"expires_at_unix_ms": storedAt + ttl, "issued_at_unix_ms": storedAt,
			"not_before_unix_ms": storedAt, "transferable": false,
		},
	}, nil
}

func buildGrantProof(trust *Trust) map[string]any {
	key := trust.keys["grant_issue"]
	issuer := cloneNode(key.principal).(map[string]any)
	issuer["authority_class"] = "forgeos_kernel"
	return map[string]any{
		"issuer": issuer, "key_id": key.id, "proof_base64url": "",
		"proof_profile_id": signatureProfile, "proof_profile_sha256": trust.profileHash,
		"trust_domain": trust.domain, "trust_epoch": trust.epoch,
	}
}

func buildSeparation(trust *Trust) map[string]any {
	return map[string]any{
		"requester":             cloneNode(trust.keys["request_auth"].principal),
		"required_distinctions": []any{"issuer_not_requester", "issuer_not_subject"},
	}
}

func buildUsagePolicy() map[string]any {
	return map[string]any{
		"atomic_reservation_required": true, "concurrent_use": "forbidden",
		"consumption_mode": "single_use", "replay": "receipt_only_no_reexecute",
		"uncertain_effect": "quarantine", "usage_ledger_required": true,
	}
}

func validateIssuanceTime(policy, request map[string]any, storedAt int64) error {
	if storedAt < 0 {
		return fmt.Errorf("durable decision time is negative")
	}
	policyValidity, _ := objectValue(policy, "validity")
	policyStart, _ := intValue(policyValidity, "not_before_unix_ms")
	policyEnd, _ := intValue(policyValidity, "expires_at_unix_ms")
	requestStart, _ := intValue(request, "requested_at_unix_ms")
	requestEnd, _ := intValue(request, "expires_at_unix_ms")
	ttl, _ := intValue(request, "requested_ttl_ms")
	if storedAt < policyStart || storedAt >= policyEnd || storedAt < requestStart || storedAt >= requestEnd {
		return fmt.Errorf("durable decision time is outside Policy or Request validity")
	}
	if policy["disposition"] == "allow" &&
		(ttl > math.MaxInt64-storedAt || storedAt+ttl > policyEnd) {
		return fmt.Errorf("Grant validity would exceed Policy validity")
	}
	return nil
}

func validateGrantRelations(grant *Grant, policy *Policy, request *Request,
	storedAt int64, trust *Trust) error {
	bytes, err := capabilitygrantcontract.CanonicalGrantJSON(grant.document)
	if err != nil || len(bytes) > maxGrantBytes {
		return fmt.Errorf("CapabilityGrant is not valid ADR-0056 canonical data")
	}
	if err = validateBootstrapGrantShape(grant.document, trust); err != nil {
		return err
	}
	if err = validateGrantRequestRelations(grant.document, policy.document, request.document); err != nil {
		return err
	}
	return validateGrantTime(grant.document, request.document, storedAt)
}

func validateBootstrapGrantShape(document map[string]any, trust *Trust) error {
	if document["issuance_phase"] != "bootstrap_planning" {
		return fmt.Errorf("Grant issuance phase is not bootstrap_planning")
	}
	approvals, err := arrayValue(document, "approval_refs")
	if err != nil || len(approvals) != 0 {
		return fmt.Errorf("bootstrap Grant approval_refs must be empty")
	}
	bindings, _ := objectValue(document, "bindings")
	if bindings["impact_sha256"] != nil || bindings["plan_sha256"] != nil || bindings["risk_sha256"] != nil {
		return fmt.Errorf("bootstrap Grant impact, plan, and risk bindings must be null")
	}
	if err := validateGrantProof(document, trust); err != nil {
		return err
	}
	return validateGrantUsageAndSeparation(document, trust)
}

func validateGrantUsageAndSeparation(document map[string]any, trust *Trust) error {
	expectedUsage := buildUsagePolicy()
	usageEqual, usageErr := sameCanonical(document["usage_policy"], expectedUsage)
	separation, separationErr := objectValue(document, "separation_of_duty")
	if usageErr != nil || !usageEqual || separationErr != nil {
		return fmt.Errorf("bootstrap Grant usage policy or separation is invalid")
	}
	expected := buildSeparation(trust)
	separationEqual, err := sameCanonical(separation, expected)
	if err != nil || !separationEqual {
		return fmt.Errorf("bootstrap Grant separation-of-duty declarations drifted")
	}
	return nil
}

func validateGrantProof(document map[string]any, trust *Trust) error {
	proof, err := objectValue(document, "authority_proof")
	if err != nil || proof["key_id"] != trust.keys["grant_issue"].id ||
		proof["proof_profile_id"] != signatureProfile || proof["proof_profile_sha256"] != trust.profileHash ||
		proof["trust_domain"] != trust.domain || proof["trust_epoch"] != trust.epoch {
		return fmt.Errorf("Grant authority proof binding is invalid")
	}
	expectedIssuer := cloneNode(trust.keys["grant_issue"].principal).(map[string]any)
	expectedIssuer["authority_class"] = "forgeos_kernel"
	equal, _ := sameCanonical(proof["issuer"], expectedIssuer)
	grantHash, _ := stringValue(document, "grant_sha256")
	proofText, _ := stringValue(proof, "proof_base64url")
	if !equal {
		return fmt.Errorf("Grant issuer does not match the pinned grant_issue principal")
	}
	return verifyDigest(trust.keys["grant_issue"].publicKey, grantSignatureDomain, grantHash, proofText)
}

func validateGrantRequestRelations(grant, policy, request map[string]any) error {
	for _, field := range []string{"budget", "capability", "scope", "subject", "task_binding"} {
		equal, err := sameCanonical(grant[field], request[field])
		if err != nil || !equal {
			return fmt.Errorf("Grant field %s differs from Request", field)
		}
	}
	bindings, _ := objectValue(grant, "bindings")
	requestBindings, _ := objectValue(request, "bindings")
	expected := map[string]any{
		"context_sha256": requestBindings["context_sha256"], "grant_request_sha256": request["request_sha256"],
		"impact_sha256": nil, "plan_sha256": nil, "policy_sha256": policy["policy_sha256"],
		"risk_sha256": nil, "source_revision": requestBindings["source_revision"],
		"source_tree_sha256": requestBindings["source_tree_sha256"],
	}
	equal, _ := sameCanonical(bindings, expected)
	if !equal {
		return fmt.Errorf("Grant bindings differ from Policy or Request")
	}
	return nil
}

func validateGrantTime(grant, request map[string]any, storedAt int64) error {
	validity, _ := objectValue(grant, "validity")
	expires, _ := intValue(validity, "expires_at_unix_ms")
	issued, _ := intValue(validity, "issued_at_unix_ms")
	notBefore, _ := intValue(validity, "not_before_unix_ms")
	ttl, _ := intValue(request, "requested_ttl_ms")
	if issued != storedAt || notBefore != storedAt || expires != storedAt+ttl || validity["transferable"] != false {
		return fmt.Errorf("Grant validity differs from durable decision time and Request TTL")
	}
	return nil
}

func grantEnvelopeSHA256(document map[string]any) (string, error) {
	return digestNode(grantEnvelopeDomain, document, maxGrantBytes, "CapabilityGrant envelope")
}

File: harness/bootstrap_grant_issuance_contract/grant.py
"""ADR-0056 CapabilityGrant overlay for the single bootstrap issuance profile."""

from __future__ import annotations

from typing import Any

from capability_grant_contract.grant import validate_grant as validate_adr0056_grant

from .authority import key_for_usage
from .canonical import ContractError, bounded_canonical_json, envelope_digest
from .constants import GRANT_ENVELOPE_DOMAIN, MAX_GRANT_BYTES, MAX_TTL_MS
from .shape import fixed_base64url, principal_identity


def grant_envelope_sha256(value: dict[str, Any]) -> str:
    return envelope_digest(GRANT_ENVELOPE_DOMAIN, value, MAX_GRANT_BYTES,
                           "complete CapabilityGrant envelope")


def validate_bootstrap_grant(value: Any, profile_hash: str,
                             root: dict[str, Any]) -> dict[str, Any]:
    node = validate_adr0056_grant(value)
    bounded_canonical_json(node, MAX_GRANT_BYTES, "bootstrap CapabilityGrant")
    if node["issuance_phase"] != "bootstrap_planning":
        raise ContractError("bootstrap Grant issuance_phase must be bootstrap_planning")
    if node["approval_refs"] != []:
        raise ContractError("bootstrap Grant approval_refs must be empty")
    bindings = node["bindings"]
    if any(bindings[field] is not None for field in ("impact_sha256", "plan_sha256",
                                                     "risk_sha256")):
        raise ContractError("bootstrap Grant impact, plan, and risk bindings must be null")
    _validate_grant_proof(node, profile_hash, root)
    _validate_usage_and_separation(node, root)
    return node


def _validate_grant_proof(node: dict[str, Any], profile_hash: str,
                          root: dict[str, Any]) -> None:
    proof = node["authority_proof"]
    issue_key = key_for_usage(root, "grant_issue")
    if proof["key_id"] != issue_key["key_id"]:
        raise ContractError("Grant proof does not use grant_issue key")
    if proof["proof_profile_id"] != "forgeos.ed25519-domain-sha256/v1":
        raise ContractError("Grant proof profile ID drifted")
    if proof["proof_profile_sha256"] != profile_hash:
        raise ContractError("Grant proof does not bind signature profile")
    if proof["trust_domain"] != root["trust_domain"] or proof["trust_epoch"] != root["trust_epoch"]:
        raise ContractError("Grant proof does not bind GovernanceTrustRoot")
    issuer = proof["issuer"]
    if issuer["authority_class"] != "forgeos_kernel":
        raise ContractError("bootstrap Grant issuer must be forgeos_kernel")
    expected = dict(issue_key["principal"])
    expected["authority_class"] = "forgeos_kernel"
    if issuer != expected:
        raise ContractError("Grant issuer differs from grant_issue principal")
    fixed_base64url(proof["proof_base64url"], "Grant proof_base64url", 64, 86)


def _validate_usage_and_separation(node: dict[str, Any], root: dict[str, Any]) -> None:
    expected_usage = {
        "atomic_reservation_required": True,
        "concurrent_use": "forbidden",
        "consumption_mode": "single_use",
        "replay": "receipt_only_no_reexecute",
        "uncertain_effect": "quarantine",
        "usage_ledger_required": True,
    }
    if node["usage_policy"] != expected_usage:
        raise ContractError("bootstrap Grant usage policy drifted")
    separation = node["separation_of_duty"]
    request_key = key_for_usage(root, "request_auth")
    expected_distinctions = ["issuer_not_requester", "issuer_not_subject"]
    if (separation["requester"] != request_key["principal"] or
            separation["required_distinctions"] != expected_distinctions):
        raise ContractError("bootstrap Grant separation-of-duty declarations drifted")
    if principal_identity(node["subject"]) == principal_identity(
            key_for_usage(root, "grant_issue")["principal"]):
        raise ContractError("grant_issue principal must differ from Grant subject")


def validate_grant_relations(grant: dict[str, Any], policy: dict[str, Any],
                             request: dict[str, Any], receipt: dict[str, Any]) -> None:
    if receipt["decision"] == "denied":
        raise ContractError("denied issuance cannot contain a Grant")
    exact = ("budget", "capability", "scope", "subject", "task_binding")
    if any(grant[field] != request[field] for field in exact):
        raise ContractError("issued Grant differs from exact Request fields")
    bindings = grant["bindings"]
    expected = {
        "context_sha256": request["bindings"]["context_sha256"],
        "grant_request_sha256": request["request_sha256"],
        "impact_sha256": None,
        "plan_sha256": None,
        "policy_sha256": policy["policy_sha256"],
        "risk_sha256": None,
        "source_revision": request["bindings"]["source_revision"],
        "source_tree_sha256": request["bindings"]["source_tree_sha256"],
    }
    if bindings != expected:
        raise ContractError("issued Grant bindings differ from Policy or Request")
    _validate_grant_validity(grant, policy, request, receipt)
    if (receipt["grant_id"] != grant["grant_id"] or
            receipt["grant_sha256"] != grant["grant_sha256"] or
            receipt["grant_envelope_sha256"] != grant_envelope_sha256(grant)):
        raise ContractError("receipt does not bind the complete issued Grant envelope")


def _validate_grant_validity(grant: dict[str, Any], policy: dict[str, Any],
                             request: dict[str, Any], receipt: dict[str, Any]) -> None:
    validity = grant["validity"]
    stored = receipt["stored_at_unix_ms"]
    if (validity["issued_at_unix_ms"] != stored or
            validity["not_before_unix_ms"] != stored or
            validity["expires_at_unix_ms"] != stored + request["requested_ttl_ms"] or
            validity["transferable"] is not False):
        raise ContractError("Grant validity does not equal durable decision time plus requested TTL")
    if validity["expires_at_unix_ms"] - validity["issued_at_unix_ms"] > MAX_TTL_MS:
        raise ContractError("Grant validity exceeds one hour")
    if validity["expires_at_unix_ms"] > policy["validity"]["expires_at_unix_ms"]:
        raise ContractError("Grant validity exceeds signed Policy validity")
